### PR TITLE
OspreySharp Phase C: collapsed stage6 worker entry path

### PR DESCRIPTION
**Pull request by brendanx67**
## Summary

Phase C of the OspreySharp pipeline-task rearchitecture: collapses the stage6 worker entry path into the canonical pipeline so the pipeline definition is the single source of truth for the four HPC-boundary tasks. The CLI mode (`--no-join --input-scores`) no longer dispatches to a separate `RescoreWorker` entry; it routes through `AnalysisPipeline.Run` with `StartAt = StopAfter = PerFileRescoreTask`, and the upstream state hydrates through `PerFileScoringTask.joinOnly`'s probe-the-disk bundle. The NextFlow consumer's contract (`re-invoke the same CLI is a no-op when outputs are valid; per-file resume on partial state`) now holds for both the in-process and worker paths from a single code path.

Five commits, each gated by build + unit tests + Stellar 3-file snapshot regression:

1. **`TaskValiditySidecar` unit tests** (`e35f928a55`). Six round-trip + edge-case tests in `OspreySharp.Test/IOTest.cs`: round-trip, JSON escapes (quotes, backslashes, `\b\f\n\r\t`, sub-0x20 control), missing-sidecar, four malformed-content shapes (truncated, missing field, garbage, empty), per-task naming collision (two tasks sharing an output path produce distinct sidecars), Delete contract (removes existing; no-op on absent). Resolves Copilot thread #3221935440 from #4199 review.

2. **`AnalysisPipeline.CanonicalPipeline()` factory** (`8363baff9b`). Pure refactor: the 4-task array construction moves out of `Run` into an internal static factory. Single source of truth for the task list ahead of the worker-collapse commit; no behavior change.

3. **Unified post-Stage-5 hydration in `PerFileScoringTask.joinOnly`** (`ad23f6356a`). Added `RescoreHydration.HydrateReconciliationOverlay(perFileEntries, parquetPaths)` so the per-file 1st-pass overlay + `reconciliation.json` parse runs from already-loaded stubs, and `RescoreHydration.HydrateForRescore` becomes a wrapper that loads stubs first. `PerFileScoringTask` owns a `RescoreInputs` bundle exposed via `GetRescoreInputs(ctx)`, populated whenever the joinOnly probe finds every parquet's `.1st-pass.fdr_scores.bin` and `reconciliation.json` siblings on disk. `FirstJoinTask`'s four reconciliation accessors fall back to the bundle when planning did not run; consensus targets are computed lazily from the post-compaction stubs.

4. **Collapsed stage6 worker entry path into the canonical pipeline** (`a20c2c1626`). `Program.Main` no longer dispatches `NoJoin+InputScores` to a separate worker; `RescoreWorker.Run` is a one-liner that calls `new AnalysisPipeline().Run(config)`. Deleted `PerFileRescoreTask.RunWorker` and its private helpers `AddIfNotNull` and `LoadOriginalRtCalibration` (-415 lines). `FirstJoinTask`'s Stage 5 work (Percolator, 1st-pass protein FDR, 1st-pass sidecar write) and Stage 6 planning gate on `bundle == null` rather than `!ExpectReconciledInput`; bundle-path compaction delegates to `RescoreCompaction.Apply` so reconciliation actions re-key to post-compaction indices. `PerFileScoringTask.joinOnly` clears the per-file PIN features after the bundle hydration so `WriteReconciledParquet`'s "rescored marker" (`Features != null`) survives. Added per-file resume inside `PerFileRescoreTask.ExecuteRescore`: files whose reconciled parquet already has a valid `.PerFileRescore.osprey.task` sidecar are skipped, and each file writes its own sidecar after `WriteReconciledParquet` so a subsequent invocation picks up only the missing files.

5. **Replaced CLI-flag dispatch with probe-the-disk on 2nd-pass sidecars** (`d577c53542`). `FirstJoinTask`'s 2nd-pass FDR overlay, `MergeNodeTask`'s 2nd-pass sidecar write gate, and `PerFileRescoreTask`'s self-gate all migrated from `ExpectReconciledInput` to a probe-the-disk check for 2nd-pass sidecar presence. `ExpectReconciledInput` survives only in `DeriveStartAtTask`/`DeriveStopAfterTask` routing and `PerFileScoringTask`'s decoy-generation skip (perf optimization that the disk state can't trivially derive). Log messages and error strings inside the 2nd-pass overlay block updated to mechanism-driven wording.

## CLI → routing mapping (no CLI surface added)

| CLI | StartAt | StopAfter | Entry path |
|-----|---------|-----------|------------|
| (default) | `PerFileScoringTask` | `MergeNodeTask` | `AnalysisPipeline.Run` |
| `--no-join` | `PerFileScoringTask` | `PerFileScoringTask` | `AnalysisPipeline.Run` |
| `--join-at-pass=1 --join-only` (no `--input-scores`) | `PerFileScoringTask` | `FirstJoinTask` | `AnalysisPipeline.Run` |
| `--join-at-pass=1 --join-only --input-scores ...` | `FirstJoinTask` | `FirstJoinTask` | `AnalysisPipeline.Run` |
| `--join-at-pass=1 --no-join --input-scores ...` | `PerFileRescoreTask` | `PerFileRescoreTask` | `AnalysisPipeline.Run` (was `RescoreWorker.Run`) |
| `--join-at-pass=2 --input-scores ...` | `MergeNodeTask` | `MergeNodeTask` | `AnalysisPipeline.Run` |

Every row routes through the same entry point. `RescoreWorker.Run` survives only as a one-line alias so external callers of the public entry-point name keep compiling.

## Test plan

- [x] Build OspreySharp solution (multi-target net472 + net8.0)
- [x] 309/309 OspreySharp unit tests pass (was 303 before this PR; +6 from commit 1)
- [x] Stellar 3-file snapshot regression PASS at every stage at every commit (`stage1to4 / stage5 / stage6 / stage7 / blib`)
- [x] Astral 3-file snapshot regression PASS at every stage
- [x] Stellar cross-impl Test-Regression PASS at every stage (Rust vs C# byte-equality)
- [x] **Manual stage6 worker crash-resume verification**: 3-file Stellar boundary inputs staged in a clean workdir.
  - Run 1: invoked `OspreySharp.exe --join-at-pass=1 --no-join --input-scores f20 --input-scores f21 --input-scores f22 -l ...` to completion. Three reconciled `.scores.parquet` files and three `.scores.parquet.PerFileRescore.osprey.task` sidecars written. Exit code 0.
  - Deleted file 21's reconciled parquet + its `.PerFileRescore.osprey.task` sidecar; restored its pre-rescore `.scores.parquet` from stage5/cs to simulate a mid-rescore crash on that file.
  - Run 2: re-invoked the same CLI. Log shows `[file] 1/3 ...20: skipping (outputs valid)`, `Re-scoring file 2/3 ...21`, `[file] 3/3 ...22: skipping (outputs valid)`. Exit code 0. All three sidecars present after Run 2.

## NextFlow integration contract

This PR is the second half of the OspreySharp HPC support work the NextFlow orchestration pipeline needs:

* **Idempotence**: re-invoking the same OspreySharp CLI on the same inputs is a no-op when the outputs (and their `.osprey.task` sidecars) are already present. Task-level skip in `AnalysisPipeline.RunTask` + per-file skip inside `PerFileScoringTask.ScoreOrLoadForFile` and `PerFileRescoreTask.ExecuteRescore` cover the two granularities.
* **Per-file resume**: per-file outputs each carry their own `<output>.<TaskName>.osprey.task` sidecar; a deleted sidecar is the orchestrator's signal "this file needs to be re-run." The orchestrator does not need to know about pipeline stages — it needs to know "which files have a valid sidecar."
* **Stable exit code semantics**: success → 0, error → non-zero (preserved from prior contract).
* **No new CLI surface added**: stage6's invocation is unchanged.

See `ai/todos/active/TODO-20260514_osprey_pipeline_tasks.md` for the design discussion, settled-decision rationale, and per-commit progress log.

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: pwiz_tools/OspreySharp/OspreySharp/Tasks/MergeNodeTask.cs [hunk 1]
```diff
@@ -112,10 +112,26 @@ public override bool Run(PipelineContext ctx)
...
+                // has a 2nd-pass sidecar on disk (replacing the prior
+                // ExpectReconciledInput gate). Under that condition the
+                // sidecar values are already correct on disk (we just
+                // loaded them in FirstJoinTask's 2nd-pass overlay), so
+                // re-writing is wasted I/O.
+                bool anyPass2OnDisk = false;
+                if (config.InputFiles != null)
+                {
+                    foreach (var inputFile in config.InputFiles)
+                    {
```
[the remark this answers is not in the document]
Comment (brendanx67) [reply pwiz_tools/OspreySharp/OspreySharp/Tasks/MergeNodeTask.cs]: Fixed in 532a83718a — switched to a per-file probe so each file's missing sidecar gets written. Restores partial-state heal on resume; the all-or-nothing gate broke that.

File: pwiz_tools/OspreySharp/OspreySharp/Tasks/FirstJoinTask.cs [hunk 2]
```diff
@@ -278,51 +338,88 @@ public override bool Run(PipelineContext ctx)
...
+            bool anyPass2Present = false;
+            if (config.InputFiles != null)
+            {
+                foreach (var inputFile in config.InputFiles)
                 {
-                    beforeCount += kvp.Value.Count;
-                    kvp.Value.RemoveAll(e => !firstPassBaseIds.Contains(e.EntryId & BASE_ID_MASK));
-                    kvp.Value.TrimExcess();
-                    afterCount += kvp.Value.Count;
+                    if (File.Exists(FdrScoresSidecar.Pass2Path(inputFile)))
```
[the remark this answers is not in the document]
Comment (brendanx67) [reply pwiz_tools/OspreySharp/OspreySharp/Tasks/FirstJoinTask.cs]: Fixed in 532a83718a — gate now rejects when a 2nd-pass binary has a stale MergeNode validity sidecar (different reconciliation/library/etc. hash). Absent sidecars are still treated as trusted boundary inputs because the snapshot test framework produces 2nd-pass binaries without sidecars (MergeNode early-exits via Environment.Exit before WriteTaskSidecars runs). Trade-off written up in the commit body.

[the remark this answers is not in the document]
Comment (brendanx67) [reply pwiz_tools/OspreySharp/OspreySharp/Tasks/FirstJoinTask.cs]: Follow-up in 2d5bd42cb4 — MergeNodeTask now writes its .MergeNode.osprey.task sidecar inline next to each 2nd-pass binary so the per-file resume contract survives OspreyDiagnostics.ExitAfterDump's Environment.Exit. Test-Snapshot.ps1 also extended (separate ai/ commit) to propagate *.osprey.task through the stage-freeze step. The relaxed gate stays as-is so Test-Regression's Rust→C# freeze (Rust never produces .osprey.task) keeps passing.
